File: src/app/services/correct-value.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { CorrectValue } from '../models/correct-value.model';
import { State } from '../models/state.model';
import { SignatureConfidence } from '../models/signature-confidence.model';

@Injectable({
  providedIn: 'root'
})
export class CorrectValueService {
  constructor(private http: HttpClient) {}

  initiateScrutinyFailure(parameterList: any): Observable<any> {
    return this.http.post('/api/scrutiny-failure', parameterList);
  }

  fetchAgents(FSC_CODE: string): Observable<any> {
    return this.http.get(`/api/agents?FSC_CODE=${FSC_CODE}`);
  }

  fetchLOVData(): Observable<any> {
    return this.http.get('/api/lov-data');
  }

  fetchAgentInfo(fscCode: string): Observable<any> {
    return this.http.get(`/api/agent-info?fscCode=${fscCode}`);
  }

  validateDate(backdatedPolicyDate: string, currentDate: Date): Observable<any> {
    return this.http.post('/api/validate-date', { backdatedPolicyDate, currentDate });
  }

  calculateAge(dateOfBirth: Date, backdatedPolicyDate: Date): Observable<any> {
    return this.http.post('/api/calculate-age', { dateOfBirth, backdatedPolicyDate });
  }

  validatePan(ageProofID: string): Observable<any> {
    return this.http.post('/api/validate-pan', { ageProofID });
  }

  saveGenderData(gender: string): Observable<any> {
    return this.http.post('/api/save-gender', { gender });
  }

  getPinCodes(): Observable<any> {
    return this.http.get('/api/pin-codes');
  }

  validateTitleGenderMaritalStatus(title: string, gender: string, maritalStatus: string): string {
    if (title === 'MRS' && gender === 'F' && maritalStatus === 'S') {
      return 'Error: Invalid combination of title, gender, and marital status.';
    }
    if (title === 'MISS' && gender === 'F' && maritalStatus === 'M') {
      return 'Error: Invalid combination of title, gender, and marital status.';
    }
    return 'Success: Valid combination.';
  }

  getStates(): Observable<State[]> {
    return this.http.get<State[]>('/api/states');
  }

  manageAgeProofField(): void {
    // Logic to manage the data and operations related to the 'IP Age Proof' field.
  }

  getCityDistrictList(country: string): Observable<any> {
    return this.http.get(`/api/cities-districts?country=${country}`);
  }

  saveResidenceProofType(proofType: string): Observable<any> {
    return this.http.post('/api/save-residence-proof', { proofType });
  }

  getResidenceProofType(): Observable<any> {
    return this.http.get('/api/get-residence-proof');
  }

  getPlacesByDistrict(district: string): Observable<any> {
    return this.http.get(`/api/places?district=${district}`);
  }

  fetchValidPinCodes(): Observable<any> {
    return this.http.get('/api/valid-pin-codes');
  }

  fetchValidAreaValues(postcode: string, state: string, district: string): Observable<any> {
    return this.http.get(`/api/valid-area-values?postcode=${postcode}&state=${state}&district=${district}`);
  }

  getLatestRequestStatus(applicationNo: string, addressType: string): Observable<any> {
    return this.http.get(`/api/latest-request-status?applicationNo=${applicationNo}&addressType=${addressType}`);
  }

  updatePanField(pan: string): Observable<any> {
    return this.http.post('/api/update-pan', { pan });
  }

  getProfessions(industryCode: string): Observable<any> {
    return this.http.get(`/api/professions?industryCode=${industryCode}`);
  }

  saveITAssesseeStatus(status: string): Observable<any> {
    return this.http.post('/api/save-it-assessee-status', { status });
  }

  saveDispatchMode(dispatchMode: string): Observable<any> {
    return this.http.post('/api/save-dispatch-mode', { dispatchMode });
  }

  getDispatchMode(): Observable<any> {
    return this.http.get('/api/get-dispatch-mode');
  }

  fetchIndustryValues(): Observable<any> {
    return this.http.get('/api/industry-values');
  }

  savePEPDetails(pepDetails: string): Observable<any> {
    return this.http.post('/api/save-pep-details', { pepDetails });
  }

  updatePEPDetails(pepDetails: string): Observable<any> {
    return this.http.post('/api/update-pep-details', { pepDetails });
  }

  fetchIncomeProofTypes(occupation: string): Observable<any> {
    return this.http.get(`/api/income-proof-types?occupation=${occupation}`);
  }

  fetchEmployerNames(): Observable<any> {
    return this.http.get('/api/employer-names');
  }

  clearFields(): void {
    // Logic to clear specific fields.
  }

  setRuralUrbanFlag(): void {
    // Logic to set the rural/urban flag based on partner type.
  }

  autoFillAadhaar(ageProofType: string, ageProofID: string): void {
    // Logic to auto-fill the Aadhaar field with the age proof ID.
  }

  queryArea(state: string, district: string, pincode: string, area: string): Observable<any> {
    return this.http.get(`/api/query-area?state=${state}&district=${district}&pincode=${pincode}&area=${area}`);
  }

  queryPlace(addressLine3: string, district: string): Observable<any> {
    return this.http.get(`/api/query-place?addressLine3=${addressLine3}&district=${district}`);
  }

  updateOccupationList(age: number): Observable<any> {
    return this.http.get(`/api/update-occupation-list?age=${age}`);
  }

  updateGender(gender: string): Observable<any> {
    return this.http.post('/api/update-gender', { gender });
  }

  saveAgeProofType(ageProofType: string): Observable<any> {
    return this.http.post('/api/save-age-proof-type', { ageProofType });
  }

  validatePAN(panID: string): Observable<any> {
    return this.http.post('/api/validate-pan', { panID });
  }

  validateAadhaar(aadhaarID: string): Observable<any> {
    return this.http.post('/api/validate-aadhaar', { aadhaarID });
  }

  getStateList(): Observable<any> {
    return this.http.get('/api/state-list');
  }

  fetchLanguages(): Observable<any> {
    return this.http.get('/api/languages');
  }

  saveSelectedLanguage(selectedLanguage: string): Observable<any> {
    return this.http.post('/api/languages', { selectedLanguage });
  }

  updateProofOfResidence(proofOfResidence: string): Observable<any> {
    return this.http.post('/api/update-proof-of-residence', { proofOfResidence });
  }

  fetchUniquePlaces(district: string): Observable<any> {
    return this.http.get(`/api/unique-places?district=${district}`);
  }

  fetchAreasAndPostalCodes(postalCode: string, state: string, district: string): Observable<any> {
    return this.http.get(`/api/areas-postal-codes?postalCode=${postalCode}&state=${state}&district=${district}`);
  }

  getCityDistrictValues(): Observable<any> {
    return this.http.get('/api/city-district-values');
  }

  fetchLatestRequestStatus(applicationNo: string, addressType: string): Observable<any> {
    return this.http.get(`/api/latest-request-status?applicationNo=${applicationNo}&addressType=${addressType}`);
  }

  fetchPinCodes(): Observable<any> {
    return this.http.get('/api/pin-codes');
  }

  fetchAreasByStateAndDistrict(state: string, district: string): Observable<any> {
    return this.http.get(`/api/areas?state=${state}&district=${district}`);
  }

  getAddressLine3Data(district: string): Observable<any> {
    return this.http.get(`/api/address-line3-data?district=${district}`);
  }

  getStatusMessage(applicationNo: string, addressType: string): Observable<any> {
    return this.http.get(`/api/status-message?applicationNo=${applicationNo}&addressType=${addressType}`);
  }

  validateMaritalStatus(maritalStatus: string, title: string, gender: string): Observable<any> {
    return this.http.post('/api/validate-marital-status', { maritalStatus, title, gender });
  }

  getProfessions(industryCode: string): Observable<any> {
    return this.http.get(`/api/professions?industryCode=${industryCode}`);
  }

  fetchValidIndustryValues(): Observable<any> {
    return this.http.get('/api/valid-industry-values');
  }

  validateIndustryDescription(industryDescription: string): Observable<any> {
    return this.http.post('/api/validate-industry-description', { industryDescription });
  }

  saveSelectedValue(selectedValue: string): Observable<any> {
    return this.http.post('/api/save-selected-value', { selectedValue });
  }

  validateJointLife(jointLifeValue: string, ip: string, ph: string): boolean {
    if (jointLifeValue === 'JL' && ip === ph) {
      return true;
    }
    return false;
  }

  saveLoanType(selectedLoanType: string): Observable<any> {
    return this.http.post('/api/save-loan-type', { selectedLoanType });
  }

  validatePortfolioStrategy(portfolioStrategy: string, frequency: string, productDefinition: string, productCode: string, packageCode: string): Observable<any> {
    return this.http.post('/api/validate-portfolio-strategy', { portfolioStrategy, frequency, productDefinition, productCode, packageCode });
  }

  incrementPhSignatureCount(): void {
    // Logic to increment the PH signature count by 1.
  }

  getBankDetails(ifscCode: string): Observable<any> {
    return this.http.get(`/api/bank-details?ifscCode=${ifscCode}`);
  }

  validateProductId(productId: string): Observable<any> {
    return this.http.post('/api/validate-product-id', { productId });
  }

  saveMoneyBackOption(value: string): Observable<any> {
    return this.http.post('/api/save-money-back-option', { value });
  }

  getMoneyBackOption(): Observable<any> {
    return this.http.get('/api/get-money-back-option');
  }

  getSignatureConfidenceDetails(applicationNumber: string): Observable<SignatureConfidence[]> {
    return this.http.get<SignatureConfidence[]>(`/api/signature-confidence/${applicationNumber}`);
  }

  validateMobileNumber(mobileNumber: string): boolean {
    const isNumeric = /^\d{10}$/.test(mobileNumber);
    return isNumeric;
  }

  validateGMI(gmiValue: number): boolean {
    return gmiValue !== null && gmiValue > 0;
  }

  calculateSumAssured(gmiValue: number): number {
    return gmiValue * 144;
  }

  validateGSTN(gstnNumber: string): boolean {
    const isValidLength = gstnNumber.length === 15;
    const hasSpecialChars = /[^a-zA-Z0-9]/.test(gstnNumber);
    return isValidLength && !hasSpecialChars;
  }

  fetchPreviousItem(currentItem: string, currentBlock: string, proposalType: string): Observable<any> {
    return this.http.get(`/api/previous-item?currentItem=${currentItem}&currentBlock=${currentBlock}&proposalType=${proposalType}`);
  }

  loadFieldData(currentItem: string, currentBlock: string, proposalType: string): Observable<any> {
    return this.http.get(`/api/load-field-data?currentItem=${currentItem}&currentBlock=${currentBlock}&proposalType=${proposalType}`);
  }

  validateAadhaar(aadhaarNumber: string, otherDetails: any): Observable<any> {
    return this.http.post('/api/validate-aadhaar', { aadhaarNumber, otherDetails });
  }

  validateSubIdCode(subIdCode: string): Observable<any> {
    return this.http.post('/api/validate-sub-id-code', { subIdCode });
  }

  fetchSurrogateIncomeProofList(): Observable<any> {
    return this.http.get('/api/surrogate-income-proof-list');
  }

  getFlagValue(): Observable<any> {
    return this.http.get('/api/flag-value');
  }

  getRateDetails(applicationNumber: string, topIndicator: string): Observable<any> {
    return this.http.get(`/api/rate-details?applicationNumber=${applicationNumber}&topIndicator=${topIndicator}`);
  }
}

File: src/main/java/com/azbj/proposalde3/service/CorrectValueService.java
package com.azbj.proposalde3.service;

import com.azbj.proposalde3.repository.CorrectValueRepository;
import com.azbj.proposalde3.model.CorrectValue;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class CorrectValueService {

    @Autowired
    private CorrectValueRepository correctValueRepository;

    public CorrectValue saveCorrectValue(CorrectValue correctValue) {
        return correctValueRepository.save(correctValue);
    }

    public Optional<CorrectValue> getCorrectValue(Long id) {
        return correctValueRepository.findById(id);
    }

    public void deleteCorrectValue(Long id) {
        correctValueRepository.deleteById(id);
    }

    public void handleScrutinyFailure(List<String> parameterList) {
        // Validate the application number
        String applicationNumber = parameterList.get(0);
        if (applicationNumber == null || applicationNumber.isEmpty()) {
            throw new IllegalArgumentException("Application number cannot be null.");
        }

        // Check user authorization
        String userId = parameterList.get(1);
        if (!userId.startsWith("UU")) {
            throw new SecurityException("Not an Authorized ID.");
        }

        // Initiate the scrutiny failure process
        // Logic to handle scrutiny failure
    }

    public List<Object[]> fetchAgents(String fscCode) {
        return correctValueRepository.findAgents(fscCode);
    }

    public List<Object[]> fetchLOVData() {
        return correctValueRepository.getLOVData();
    }

    public boolean validateDate(String backdatedPolicyDate, java.util.Date currentDate) {
        // Logic to validate the backdated policy date
        return !backdatedPolicyDate.after(currentDate.toString());
    }

    public int calculateAge(java.util.Date dateOfBirth, java.util.Date backdatedPolicyDate) {
        // Logic to calculate age
        return backdatedPolicyDate.getYear() - dateOfBirth.getYear();
    }

    public boolean validatePan(String ageProofID) {
        // Logic to validate PAN
        return ageProofID != null && ageProofID.matches("[A-Z]{5}[0-9]{4}[A-Z]{1}");
    }

    public void saveGenderData(String gender) {
        // Logic to save gender data
        CorrectValue correctValue = new CorrectValue();
        correctValue.setGender(gender);
        correctValueRepository.save(correctValue);
    }

    public List<Object[]> getDistinctPinCodes() {
        return correctValueRepository.findDistinctPinCodes();
    }

    public List<Object[]> fetchStates() {
        return correctValueRepository.findAllStates();
    }

    public List<Object[]> fetchAddressDetails(String landmarkArea) {
        return correctValueRepository.findAddressDetails(landmarkArea);
    }

    public List<Object[]> fetchLanguages() {
        return correctValueRepository.findAll();
    }

    public void storeLanguage(String selectedLanguage) {
        CorrectValue correctValue = new CorrectValue();
        correctValue.setLanguage(selectedLanguage);
        correctValueRepository.save(correctValue);
    }

    public String getRequestStatus(String applicationNo) {
        return correctValueRepository.getRequestStatus(applicationNo);
    }

    public List<Object[]> getPlacesByDistrict(String district) {
        return correctValueRepository.findUniquePlacesByDistrict(district);
    }

    public List<Object[]> fetchValidPinCodes() {
        return correctValueRepository.findValidPinCodes();
    }

    public List<Object[]> fetchValidAreaValues(String postcode, String state, String district) {
        return correctValueRepository.findValidAreaValues(postcode, state, district);
    }

    public String getLatestRequestStatus(String applicationNo, String addressType) {
        return correctValueRepository.findLatestRequestStatus(applicationNo, addressType);
    }

    public void handlePanField(String pan) {
        CorrectValue correctValue = new CorrectValue();
        correctValue.setPan(pan);
        correctValueRepository.save(correctValue);
    }

    public List<Object[]> getProfessions(String industryCode) {
        return correctValueRepository.findProfessionsByIndustryCode(industryCode);
    }

    public List<Object[]> fetchIndustryValues() {
        return correctValueRepository.getIndustryValues();
    }

    public boolean validateIndustryDescription(String industryDescription, String appln_no) {
        int count = correctValueRepository.countRelevantRecords(appln_no);
        return count > 0 && (industryDescription.equals("CRPF") || industryDescription.equals("DEFENCE"));
    }

    public void savePEPDetails(String pepDetails) {
        CorrectValue correctValue = new CorrectValue();
        correctValue.setPepDetails(pepDetails);
        correctValueRepository.save(correctValue);
    }

    public void updatePEPDetails(String pepDetails) {
        CorrectValue correctValue = new CorrectValue();
        correctValue.setPepDetails(pepDetails);
        correctValueRepository.save(correctValue);
    }

    public List<Object[]> fetchIncomeProofTypes(String occupation) {
        return correctValueRepository.findIncomeProofTypesByOccupation(occupation);
    }

    public List<Object[]> fetchEmployerNames() {
        return correctValueRepository.findEmployerNames();
    }

    public List<Object[]> queryArea(String state, String district, String pincode, String area) {
        return correctValueRepository.queryArea(state, district, pincode, area);
    }

    public List<Object[]> queryPlace(String addressLine3, String district) {
        return correctValueRepository.queryPlace(addressLine3, district);
    }

    public List<Object[]> updateOccupationList(int age) {
        return correctValueRepository.updateOccupationList(age);
    }

    public boolean validatePAN(String panID) {
        return panID != null && panID.matches("[A-Z]{5}[0-9]{4}[A-Z]{1}");
    }

    public boolean validateAadhaar(String aadhaarID) {
        return aadhaarID != null && aadhaarID.matches("[0-9]{12}");
    }

    public List<Object[]> getPinCodes() {
        return correctValueRepository.findDistinctPinCodes();
    }

    public List<Object[]> getStateList() {
        return correctValueRepository.fetchStateList();
    }

    public List<Object[]> getLanguages() {
        return correctValueRepository.findAll();
    }

    public void saveSelectedLanguage(String selectedLanguage) {
        CorrectValue correctValue = new CorrectValue();
        correctValue.setLanguage(selectedLanguage);
        correctValueRepository.save(correctValue);
    }

    public List<Object[]> fetchUniquePlaces(String district) {
        return correctValueRepository.findUniquePlaces(district);
    }

    public List<Object[]> fetchAreasAndPostalCodes(String postalCode, String state, String district) {
        return correctValueRepository.fetchAreasAndPostalCodes(postalCode, state, district);
    }

    public String fetchLatestRequestStatus(String applicationNo, String addressType) {
        return correctValueRepository.findLatestRequestStatus(applicationNo, addressType);
    }

    public List<Object[]> fetchAreasByStateAndDistrict(String state, String district) {
        return correctValueRepository.findAreasByStateAndDistrict(state, district);
    }

    public List<Object[]> getAddressLine3Data(String district) {
        return correctValueRepository.fetchAddressLine3Data(district);
    }

    public String getStatusMessage(String applicationNo, String addressType) {
        return correctValueRepository.fetchStatusMessage(applicationNo, addressType);
    }

    public List<Object[]> fetchProfessions(String industryCode) {
        return correctValueRepository.getProfessions(industryCode);
    }

    public boolean validateIndustryDescription(String industryDescription, String appln_no, String fsc_code, String v_inhouse_flag, String v_web_aggr) {
        int count = correctValueRepository.countRecords(appln_no, fsc_code, v_inhouse_flag, v_web_aggr);
        return count > 0 && (industryDescription.equals("CRPF") || industryDescription.equals("DEFENCE"));
    }

    public void saveSelectedValue(String selectedValue) {
        CorrectValue correctValue = new CorrectValue();
        correctValue.setSelectedValue(selectedValue);
        correctValueRepository.save(correctValue);
    }

    public void saveLoanType(String selectedLoanType) {
        CorrectValue correctValue = new CorrectValue();
        correctValue.setLoanType(selectedLoanType);
        correctValueRepository.save(correctValue);
    }

    public boolean validatePortfolioStrategy(String portfolioStrategy, String frequency, String productDefinition, String productCode, String packageCode) {
        // Logic to validate portfolio strategy
        return true;
    }

    public List<Object[]> fetchBankDetails(String ifscCode) {
        return correctValueRepository.findByIfscCode(ifscCode);
    }

    public boolean validateProductId(String productId) {
        return correctValueRepository.findProductById(productId).isPresent();
    }

    public void saveMoneyBackOption(String value) {
        CorrectValue correctValue = new CorrectValue();
        correctValue.setMoneyBackOption(value);
        correctValueRepository.save(correctValue);
    }

    public String getMoneyBackOption() {
        return correctValueRepository.findMoneyBackOption();
    }

    public List<Object[]> getSignatureConfidenceDetails(String applicationNumber) {
        return correctValueRepository.findByApplicationNumber(applicationNumber);
    }

    public boolean validateMobileNumber(String mobileNumber) {
        return mobileNumber != null && mobileNumber.matches("[0-9]{10}");
    }

    public boolean validateGMI(String gmiValue) {
        return gmiValue != null && !gmiValue.isEmpty();
    }

    public int calculateSumAssured(String gmiValue) {
        return Integer.parseInt(gmiValue) * 144;
    }

    public String fetchPreviousItem(String currentItem, String currentBlock, String proposalType) {
        return correctValueRepository.fetchPreviousItem(currentItem, currentBlock, proposalType);
    }

    public String loadFieldData(String currentItem, String currentBlock, String proposalType) {
        return correctValueRepository.loadFieldData(currentItem, currentBlock, proposalType);
    }

    public String validateChannel(String applicationNo, String enteredChannel) {
        String channel = correctValueRepository.findChannelInBatchItems(applicationNo);
        if (channel == null) {
            channel = correctValueRepository.findChannelInScrutinyPropExtn(applicationNo);
        }
        return channel.equals(enteredChannel) ? "green" : "red";
    }

    public boolean validateAadhaar(String aadhaarNumber, String otherDetails) {
        return aadhaarNumber != null && aadhaarNumber.matches("[0-9]{12}");
    }

    public boolean validateSubIdCode(String subIdCode) {
        return correctValueRepository.validateSubIdCode(subIdCode);
    }

    public List<Object[]> fetchRateDetails(String applicationNumber, String topIndicator) {
        return correctValueRepository.findRateDetails(applicationNumber, topIndicator);
    }
}